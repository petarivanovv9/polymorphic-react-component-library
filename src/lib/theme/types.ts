import type { BadgeThemeProps } from './badge/types';
import type { ButtonThemeProps } from './button/types';

export interface ExampleUITheme extends Record<string, unknown> {
  badge: BadgeThemeProps;
  button: ButtonThemeProps;
}
