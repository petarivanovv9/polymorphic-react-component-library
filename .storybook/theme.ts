import { version } from '../package.json';

export const theme = {
    light: {},
    dark: {},
};
