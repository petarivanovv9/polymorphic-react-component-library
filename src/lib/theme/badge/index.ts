import type { BadgeThemeProps } from './types';

export const badge: BadgeThemeProps = {
  base: `
        inline-flex align-center items-center align-self-center gap-1
        rounded-full border border-transparent font-semibold
    `,
  variant: {
    primary: 'text-primary-500 bg-primary-50 border-primary-300',
    secondary: 'text-secondary-500 bg-secondary-50 border-secondary-300',
    success: 'text-success-500 bg-success-50 border-success-300',
    danger: 'text-danger-500 bg-danger-50 border-danger-300',
    warning: 'text-warning-500 bg-warning-50 border-warning-300',
    info: 'text-info-500 bg-info-50 border-info-300',
  },
  size: {
    small: 'text-xs min-h-5.5 px-1 py-0.5',
    medium: 'text-sm min-h-7.5 px-2 py-1',
    large: 'text-base min-h-9.5 px-3 py-1.5',
    'extra-large': 'text-lg min-h-11.5 px-4 py-2',
  },
};
