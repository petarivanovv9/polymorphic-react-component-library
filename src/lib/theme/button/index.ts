import type { ButtonThemeProps } from './types';

export const button: ButtonThemeProps = {
  base: `
        appearance-none cursor-pointer 
        inline-flex align-center items-center align-self-center gap-2
        transition duration-200 ease
        border border-transparent rounded-md
        hover:shadow-md
        focus:outline-none focus:ring-2
        disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none
    `,
  variant: {
    // Solid Variants
    primary: `
            text-white bg-primary-500 ring-primary-300
            hover:bg-primary-700
            active:bg-primary-900
            disabled:hover:bg-primary-500 
        `,
    secondary: `
            text-white bg-secondary-500 ring-secondary-300
            hover:bg-secondary-700
            active:bg-secondary-900
            disabled:hover:bg-secondary-500
        `,
    success: `
            text-white bg-success-500 ring-success-300
            hover:bg-success-700
            active:bg-success-900
            disabled:hover:bg-success-500
        `,
    danger: `
            text-white bg-danger-500 ring-danger-300
            hover:bg-danger-700
            active:bg-danger-900
            disabled:hover:bg-danger-500
        `,
    warning: `
            text-white bg-warning-500 ring-warning-300
            hover:bg-warning-700
            active:bg-warning-900
            disabled:hover:bg-warning-500
        `,
    info: `
            text-white bg-info-500 ring-info-300 
            hover:bg-info-700
            active:bg-info-900
            disabled:hover:bg-info-500
        `,
    // Outline Variants
    'outline-primary': `
            text-primary-500 bg-white border-primary-500 ring-primary-300
            hover:text-primary-700 hover:bg-primary-50 hover:border-primary-700
            active:border-primary-800 active:bg-primary-100
            disabled:hover:text-primary-500 disabled:hover:bg-white disabled:hover:border-primary-500
        `,
    'outline-secondary': `
            text-secondary-500 bg-white border-secondary-500 ring-secondary-300
            hover:text-secondary-700 hover:bg-secondary-50 hover:border-secondary-700
            active:border-secondary-800 active:bg-secondary-100
            disabled:hover:text-secondary-500 disabled:hover:bg-white disabled:hover:border-secondary-500
        `,
    'outline-success': `
            text-success-500 bg-white border-success-500 ring-success-300
            hover:text-success-700 hover:bg-success-50 hover:border-success-700
            active:border-success-800 active:bg-success-100
            disabled:hover:text-success-500 disabled:hover:bg-white disabled:hover:border-success-500
        `,
    'outline-danger': `
            text-danger-500 bg-white border-danger-500 ring-danger-300
            hover:text-danger-700 hover:bg-danger-50 hover:border-danger-700
            active:border-danger-800 active:bg-danger-100
            disabled:hover:text-danger-500 disabled:hover:bg-white disabled:hover:border-danger-500
        `,
    'outline-warning': `
            text-warning-500 bg-white border-warning-500 ring-warning-300
            hover:text-warning-700 hover:bg-warning-50 hover:border-warning-700
            active:border-warning-800 active:bg-warning-100
            disabled:hover:text-warning-500 disabled:hover:bg-white disabled:hover:border-warning-500
        `,
    'outline-info': `
            text-info-500 bg-white border-info-500 ring-info-300
            hover:text-info-700 hover:bg-info-50 hover:border-info-700
            active:border-info-800 hover:bg-info-100
            disabled:hover:text-info-500 disabled:hover:bg-white disabled:hover:border-info-500
        `,
    // Ghost Variants
    'ghost-primary': `
            text-primary-500 bg-white border-transparent ring-primary-300
            hover:text-primary-700 hover:bg-primary-50 hover:border-primary-700
            active:border-primary-800 active:bg-primary-100
            disabled:hover:text-primary-500 disabled:hover:bg-white disabled:hover:border-transparent
        `,
    'ghost-secondary': `
            text-secondary-500 bg-white border-transparent ring-secondary-300
            hover:text-secondary-700 hover:bg-secondary-50 hover:border-secondary-700
            active:border-secondary-800 active:bg-secondary-100
            disabled:hover:text-secondary-500 disabled:hover:bg-white disabled:hover:border-transparent
        `,
    'ghost-success': `
            text-success-500 bg-white border-transparent ring-success-300
            hover:text-success-700 hover:bg-success-50 hover:border-success-700
            active:border-success-800 active:bg-success-100
            disabled:hover:text-success-500 disabled:hover:bg-white disabled:hover:border-transparent
        `,
    'ghost-danger': `
            text-danger-500 bg-white border-transparent ring-danger-300
            hover:text-danger-700 hover:bg-danger-50 hover:border-danger-700
            active:border-danger-800 active:bg-danger-100
            disabled:hover:text-danger-500 disabled:hover:bg-white disabled:hover:border-transparent
        `,
    'ghost-warning': `
            text-warning-500 bg-white border-transparent ring-warning-300
            hover:text-warning-700 hover:bg-warning-50 hover:border-warning-700
            active:border-warning-800 active:bg-warning-100
            disabled:hover:text-warning-500 disabled:hover:bg-white disabled:hover:border-transparent
        `,
    'ghost-info': `
            text-info-500 bg-white border-transparent ring-info-300
            hover:text-info-700 hover:bg-info-50 hover:border-info-700
            active:border-info-800 hover:bg-info-100
            disabled:hover:text-info-500 disabled:hover:bg-white disabled:hover:border-transparent
        `,
    // Link Variants
    'link-primary': `
            text-primary-500 bg-transparent border-transparent ring-primary-300
            hover:text-primary-700 hover:transparent hover:underline hover:shadow-none
            disabled:hover:text-primary-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline
        `,
    'link-secondary': `
            text-secondary-500 bg-transparent border-transparent ring-secondary-300
            hover:text-secondary-700 hover:transparent hover:underline hover:shadow-none
            disabled:hover:text-secondary-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline
        `,
    'link-success': `
            text-success-500 bg-transparent border-transparent ring-success-300
            hover:text-success-700 hover:transparent hover:underline hover:shadow-none
            disabled:hover:text-success-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline
        `,
    'link-danger': `
            text-danger-500 bg-transparent border-transparent ring-danger-300
            hover:text-danger-700 hover:transparent hover:underline hover:shadow-none
            disabled:hover:text-danger-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline
        `,
    'link-warning': `
            text-warning-500 bg-transparent border-transparent ring-warning-300
            hover:text-warning-700 hover:transparent hover:underline hover:shadow-none
            disabled:hover:text-warning-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline
        `,
    'link-info': `
            text-info-500 bg-transparent border-transparent ring-info-300
            hover:text-info-700 hover:transparent hover:underline hover:shadow-none
            disabled:hover:text-info-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline
        `,
  },
  size: {
    'extra-small': 'text-xs px-1 py-0.5 min-h-5.5',
    small: 'text-sm px-2 py-1 min-h-7.5',
    medium: 'text-base px-3 py-1.5 min-h-9.5',
    large: 'text-lg px-4 py-2 min-h-11.5',
    'extra-large': 'text-xl px-5 py-3 min-h-13.5',
  },
  fullWidth: `w-full justify-center`,
};
