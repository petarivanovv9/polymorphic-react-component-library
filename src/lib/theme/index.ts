import type { ExampleUITheme } from './types';

import { badge } from './badge';
import { button } from './button';

export type { ExampleUITheme };
export const theme: ExampleUITheme = {
  badge,
  button,
};
