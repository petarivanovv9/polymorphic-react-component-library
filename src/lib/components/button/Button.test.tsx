import { describe, expect, it } from 'vitest';

import { Button } from '..';
import {
  render,
  renderWithThemeProvider,
  screen,
} from '../../../utils/test-utils';
import { theme as defaultTheme } from '../../theme';
import type { ExampleUITheme } from '../../theme/types';

const defaultProps = {
  children: 'Content',
};

const renderComponent = (props = {}) =>
  render(<Button {...defaultProps} {...props} />);

const renderComponentWithThemeProvider = (props = {}, context = {}) =>
  renderWithThemeProvider(<Button {...defaultProps} {...props} />, context);

describe('Button', () => {
  describe('Props', () => {
    it('should match snapshot', () => {
      const { container } = renderComponent();

      expect(container).toMatchSnapshot();
      expect(screen.getByRole('button')).toHaveAttribute('type', 'button');
    });

    it('should match snapshot when rendering `as` anchor `<a>`', () => {
      const props = {
        as: 'a',
        href: 'https://random.com',
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
      expect(screen.getByRole('link')).toBeInTheDocument();
      expect(screen.getByRole('link')).not.toHaveAttribute('type');
    });

    it('should match snapshot when rendering `as` React component', () => {
      const Link = () => <a href="https://random.com">Link</a>;
      const props = {
        as: Link,
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
      expect(screen.getByRole('link')).toBeInTheDocument();
      expect(screen.getByRole('link')).not.toHaveAttribute('type');
    });

    it('should match snapshot with `size`', () => {
      const props = {
        size: 'large',
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with default (solid) `variant`', () => {
      const props = {
        variant: 'primary',
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with outline `variant`', () => {
      const props = {
        variant: 'outline-primary',
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with ghost `variant`', () => {
      const props = {
        variant: 'ghost-primary',
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with link `variant`', () => {
      const props = {
        variant: 'link-primary',
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with `fullWidth`', () => {
      const props = {
        fullWidth: true,
      };
      const { container } = renderComponent(props);

      expect(container).toMatchSnapshot();
    });

    it('should extend styles with custom `className`', () => {
      const className = 'bg-red-500';
      const props = {
        className,
      };
      renderComponent(props);

      expect(screen.getByRole('button')).toHaveClass(className);
    });

    it('should be disabled when `disabled={true}`', () => {
      const props = {
        disabled: true,
      };
      renderComponent(props);

      expect(screen.getByRole('button')).toBeDisabled();
    });
  });

  describe('Accessibility', () => {
    it('should have `role="button"` by default', () => {
      renderComponent();

      expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('should have `role="link"` when rendering `as` an anchor `<a>`', () => {
      const props = {
        as: 'a',
        href: 'https://random.com',
      };
      renderComponent(props);

      expect(screen.getByRole('link')).toBeInTheDocument();
    });
  });

  describe('Theming', () => {
    it('should use `base` classes', () => {
      const classes = 'outline-none';
      const customTheme: ExampleUITheme = {
        ...defaultTheme,
        button: {
          ...defaultTheme.button,
          base: classes,
        },
      };
      const context = {
        theme: customTheme,
      };

      renderComponentWithThemeProvider(undefined, context);

      expect(screen.getByRole('button')).toHaveClass(classes);
    });

    it('should use solid `variant` classes', () => {
      const classes = 'bg-pink-500';
      const customTheme: ExampleUITheme = {
        ...defaultTheme,
        button: {
          ...defaultTheme.button,
          variant: {
            ...defaultTheme.button.variant,
            primary: classes,
          },
        },
      };

      const props = {
        variant: 'primary',
      };
      const context = {
        theme: customTheme,
      };

      renderComponentWithThemeProvider(props, context);

      expect(screen.getByRole('button')).toHaveClass(classes);
    });

    it('should use outline `variant` classes', () => {
      const classes = 'bg-pink-500';
      const customTheme: ExampleUITheme = {
        ...defaultTheme,
        button: {
          ...defaultTheme.button,
          variant: {
            ...defaultTheme.button.variant,
            'outline-primary': classes,
          },
        },
      };
      const props = {
        variant: 'outline-primary',
      };
      const context = {
        theme: customTheme,
      };

      renderComponentWithThemeProvider(props, context);

      expect(screen.getByRole('button')).toHaveClass(classes);
    });

    it('should use ghost `variant` classes', () => {
      const classes = 'bg-pink-500';
      const customTheme: ExampleUITheme = {
        ...defaultTheme,
        button: {
          ...defaultTheme.button,
          variant: {
            ...defaultTheme.button.variant,
            'ghost-primary': classes,
          },
        },
      };
      const props = {
        variant: 'ghost-primary',
      };
      const context = {
        theme: customTheme,
      };

      renderComponentWithThemeProvider(props, context);

      expect(screen.getByRole('button')).toHaveClass(classes);
    });

    it('should use link `variant` classes', () => {
      const classes = 'bg-pink-500';
      const customTheme: ExampleUITheme = {
        ...defaultTheme,
        button: {
          ...defaultTheme.button,
          variant: {
            ...defaultTheme.button.variant,
            'link-primary': classes,
          },
        },
      };
      const props = {
        variant: 'link-primary',
      };
      const context = {
        theme: customTheme,
      };

      renderComponentWithThemeProvider(props, context);

      expect(screen.getByRole('button')).toHaveClass(classes);
    });

    it('should use `size` classes', () => {
      const classes = 'font-extralight';
      const customTheme: ExampleUITheme = {
        ...defaultTheme,
        button: {
          ...defaultTheme.button,
          size: {
            ...defaultTheme.button.size,
            small: classes,
          },
        },
      };
      const props = {
        size: 'small',
      };
      const context = {
        theme: customTheme,
      };

      renderComponentWithThemeProvider(props, context);

      expect(screen.getByRole('button')).toHaveClass(classes);
    });
  });
});
