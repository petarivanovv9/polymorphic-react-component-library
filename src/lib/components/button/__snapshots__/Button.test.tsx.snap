// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > Props > should match snapshot 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border border-transparent rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-white bg-secondary-500 ring-secondary-300 hover:bg-secondary-700 active:bg-secondary-900 disabled:hover:bg-secondary-500 text-base px-3 py-1.5 min-h-9.5"
    type="button"
  >
    Content
  </button>
</div>
`;

exports[`Button > Props > should match snapshot when rendering \`as\` React component 1`] = `
<div>
  <a
    href="https://random.com"
  >
    Link
  </a>
</div>
`;

exports[`Button > Props > should match snapshot when rendering \`as\` anchor \`<a>\` 1`] = `
<div>
  <a
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border border-transparent rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-white bg-secondary-500 ring-secondary-300 hover:bg-secondary-700 active:bg-secondary-900 disabled:hover:bg-secondary-500 text-base px-3 py-1.5 min-h-9.5"
    href="https://random.com"
  >
    Content
  </a>
</div>
`;

exports[`Button > Props > should match snapshot with \`fullWidth\` 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border border-transparent rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-white bg-secondary-500 ring-secondary-300 hover:bg-secondary-700 active:bg-secondary-900 disabled:hover:bg-secondary-500 text-base px-3 py-1.5 min-h-9.5 w-full justify-center"
    type="button"
  >
    Content
  </button>
</div>
`;

exports[`Button > Props > should match snapshot with \`size\` 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border border-transparent rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-white bg-secondary-500 ring-secondary-300 hover:bg-secondary-700 active:bg-secondary-900 disabled:hover:bg-secondary-500 text-lg px-4 py-2 min-h-11.5"
    type="button"
  >
    Content
  </button>
</div>
`;

exports[`Button > Props > should match snapshot with default (solid) \`variant\` 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border border-transparent rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-white bg-primary-500 ring-primary-300 hover:bg-primary-700 active:bg-primary-900 disabled:hover:bg-primary-500 text-base px-3 py-1.5 min-h-9.5"
    type="button"
  >
    Content
  </button>
</div>
`;

exports[`Button > Props > should match snapshot with ghost \`variant\` 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-primary-500 bg-white border-transparent ring-primary-300 hover:text-primary-700 hover:bg-primary-50 hover:border-primary-700 active:border-primary-800 active:bg-primary-100 disabled:hover:text-primary-500 disabled:hover:bg-white disabled:hover:border-transparent text-base px-3 py-1.5 min-h-9.5"
    type="button"
  >
    Content
  </button>
</div>
`;

exports[`Button > Props > should match snapshot with link \`variant\` 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border rounded-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-primary-500 bg-transparent border-transparent ring-primary-300 hover:text-primary-700 hover:transparent hover:underline hover:shadow-none disabled:hover:text-primary-500 disabled:hover:bg-transparent disabled:hover:border-transparent disabled:hover:no-underline text-base px-3 py-1.5 min-h-9.5"
    type="button"
  >
    Content
  </button>
</div>
`;

exports[`Button > Props > should match snapshot with outline \`variant\` 1`] = `
<div>
  <button
    class="appearance-none cursor-pointer inline-flex align-center items-center align-self-center gap-2 transition duration-200 ease border rounded-md hover:shadow-md focus:outline-none focus:ring-2 disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none text-primary-500 bg-white border-primary-500 ring-primary-300 hover:text-primary-700 hover:bg-primary-50 hover:border-primary-700 active:border-primary-800 active:bg-primary-100 disabled:hover:text-primary-500 disabled:hover:bg-white disabled:hover:border-primary-500 text-base px-3 py-1.5 min-h-9.5"
    type="button"
  >
    Content
  </button>
</div>
`;
